internal/util/sessionutil/session_util.go

```diff
@@ -598,3 +607,99 @@ func (s *Session) Registered() bool {
 	}
 	return b
 }
+
+func (s *Session) SetEnableActiveStandBy(enable bool) {
+	s.enableActiveStandBy = enable
+}
+
+func (s *Session) updateStandby(b bool) {
+	s.isStandby.Store(b)
+}
+
+// ProcessActiveStandBy is used by coordinators to do active-standby mechanism.
+// coordinator enabled active-standby will first call Register and then call ProcessActiveStandBy.
+// steps:
+// 1, Enter STANDBY mode
+// 2, Try to register to active key.
+// 3, If 2. succeeds, this service becomes ACTIVE. Exit STANDBY mode.
+// 4, If 2. fails after retrying, which means an ACTIVE service already exist.
+//    it will keep watching the active key. Whenever active key disappears, STANDBY node will go backup to 2.
+// activateFunc is the function to re-active the service.
+func (s *Session) ProcessActiveStandBy(activateFunc func()) error {
+	s.activeKey = path.Join(s.metaRoot, DefaultServiceRoot, s.ServerName)
+
+	registerActiveFn := func() error {
+		sessionJSON, err := json.Marshal(s)
+		if err != nil {
+			return err
+		}
+		txnResp, err := s.etcdCli.Txn(s.ctx).If(
+			clientv3.Compare(
+				clientv3.Version(s.activeKey),
+				"=",
+				0)).
+			Then(clientv3.OpPut(s.activeKey, string(sessionJSON), clientv3.WithLease(*s.leaseID))).Commit()
+		if err != nil {
+			log.Warn("compare and swap error, maybe the key has already been registered", zap.Error(err))
+			return err
+		}
+		if !txnResp.Succeeded {
+			return fmt.Errorf("function CompareAndSwap error for compare is false for key: %s", s.activeKey)
+		}
+		log.Info("register ACTIVE service successfully",
+			zap.String("ServerName", s.ServerName),
+			zap.Int64("ServerID", s.ServerID))
+		return nil
+	}
+
+	s.updateStandby(true)
+	log.Info(fmt.Sprintf("serverName: %v enter STANDBY mode", s.ServerName))
+	go func() {
+		for s.isStandby.Load().(bool) {
+			log.Info(fmt.Sprintf("serverName: %v is in STANDBY ...", s.ServerName))
+			time.Sleep(10 * time.Second)
+		}
+	}()
+
+	for {
+		log.Info("start retrying to register as ACTIVE service...")
+		err := retry.Do(s.ctx, registerActiveFn, retry.Attempts(uint(GlobalParams.CommonCfg.SessionRetryTimes)), retry.Sleep(100*time.Millisecond))
+		if err != nil {
+			// watch the active key in etcd
+			log.Info("try to register as ACTIVE service fail after retry, there should be other ACTIVE service. Start to watch ACTIVE key")
+			ctx, cancel := context.WithCancel(s.ctx)
+			watchChan := s.etcdCli.Watch(ctx, s.activeKey, clientv3.WithPrefix(), clientv3.WithPrevKV())
```

Comment:
edged case:
If you read there is a kv, but watch happened after the kv expire, then you never know the master change.

Some suggestion:
1) registerActiveFn should return succeed/fail rather than compare and swap , but not error if compare and swap
2) registerActiveFn should return revision if there exist a current leader
3) etcd watch the specified revision to avoid missing some notification